packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts

```diff
@@ -576,22 +574,32 @@ export class OdspDocumentDeltaConnection extends DocumentDeltaConnection {
         }
     }
 
+    /**
+     * Critical path where we need to also close the socket for an error.
+     * @param error - Error causing the socket to close.
+     */
+    protected disposeSocket(error: IAnyDriverError) {
+        const socket = this.socketReference;
+        assert(socket !== undefined, "reentrancy not supported in close socket");
+        socket.closeSocket();
+        this.disposeCore(error);
```

Comment:
But why does this client needs to emit any event. One thing is that socket.disconnect() implicitly raises an event on the socket. But for example, when the server will raise serverDisconnect all the client will receive that message/event and do the needful. We just need to give them the chance to listen to the correct event and doing socket close on next js turn.

I do however find some benefit of raising an event with some common name and with a reason, and then other client will get that reason instead of implicit reason and every client will listen to this event and will close socket. Is this what you are suggesting?

Reply:
I think your point - every connection should get here when socket is closed, and thus they already disconnecting all connections. I think that's a valid point, assuming we do not have any corner cases or race conditions, where client did not have a chance to register (though in such case it will not have time to register for this event either).

I'd say these two changes are needed:
1. First, I think lines 584 & 585 should be reordered. Otherwise, we may lose disconnect reason (we do not today due to closeSocket closing it in async task, but if that were to change, things would be bad). I.e. connections should have a chance to detach from socket before we move to closing socket.
2. We should probably log error in closeSocket() that we do not have any connections on the socket when it's actually being closed. I think if we do it synchronously (with # 1 above done), then likely stack will help us figure out where violation happened (in this case violation will mean - we are losing reason for disconnect).

Reply:
Did the first thing. In second are you suggesting making the socket close synchronous on line 158 instead on next js turn?

Reply:
Added an assert to check references should be 0 to socket at this point.